OpenRA.Mods.Common/Traits/Upgrades/UpgradeOnDamageState.cs

```diff
@@ -32,41 +32,55 @@ public class UpgradeOnDamageInfo : ITraitInfo, Requires<UpgradeManagerInfo>
 		[Desc("Are upgrades irrevocable once the conditions have been met?")]
 		public readonly bool GrantPermanently = false;
 
-		public object Create(ActorInitializer init) { return new UpgradeOnDamage(init.Self, this); }
+		public object Create(ActorInitializer init) { return new UpgradeOnDamageState(init.Self, this); }
 	}
 
-	public class UpgradeOnDamage : INotifyDamageStateChanged
+	public class UpgradeOnDamageState : INotifyDamageStateChanged, INotifyCreated
 	{
-		readonly UpgradeOnDamageInfo info;
+		readonly UpgradeOnDamageStateInfo info;
 		readonly UpgradeManager um;
+		readonly Health health;
 		bool granted;
 
-		public UpgradeOnDamage(Actor self, UpgradeOnDamageInfo info)
+		public UpgradeOnDamageState(Actor self, UpgradeOnDamageStateInfo info)
 		{
 			this.info = info;
-			um = self.TraitOrDefault<UpgradeManager>();
+			um = self.Trait<UpgradeManager>();
+			health = self.Trait<Health>();
 		}
 
-		void INotifyDamageStateChanged.DamageStateChanged(Actor self, AttackInfo e)
+		void INotifyCreated.Created(Actor self)
 		{
-			if (um == null)
-				return;
-
-			if (granted && info.GrantPermanently)
-				return;
+			if (!granted)
+				GrantUpgradeOnValidDamageState(self, health.DamageState);
+		}
 
-			var rand = Game.CosmeticRandom;
-			if (!granted && info.ValidDamageStates.HasFlag(e.DamageState) && !info.ValidDamageStates.HasFlag(e.PreviousDamageState))
+		void GrantUpgradeOnValidDamageState(Actor self, DamageState state)
+		{
+			if (info.ValidDamageStates.HasFlag(state))
 			{
 				granted = true;
+				var rand = Game.CosmeticRandom;
 				var sound = info.EnabledSounds.RandomOrDefault(rand);
 				Game.Sound.Play(sound, self.CenterPosition);
 				foreach (var u in info.Upgrades)
 					um.GrantUpgrade(self, u, this);
 			}
+		}
+
+		void INotifyDamageStateChanged.DamageStateChanged(Actor self, AttackInfo e)
+		{
+			if (granted && info.GrantPermanently)
+				return;
+
+			if (!granted && !info.ValidDamageStates.HasFlag(e.PreviousDamageState))
+			{
+				GrantUpgradeOnValidDamageState(self, health.DamageState);
+			}
```

Comment:
Well I'll leave the final say to @pchote but I have never had these brackets even with the else if,
except when there is an if inside that if.